packages/std/src/stdack.rs

```diff
@@ -0,0 +1,163 @@
+    Success(Binary),
+    Error(String),
+}
+
+impl StdAck {
+    /// Creates a success ack with the given data
+    pub fn success(data: impl Into<Binary>) -> Self {
+        StdAck::Success(data.into())
+    }
+
+    /// Creates an error ack
+    pub fn error(err: impl Into<String>) -> Self {
+        StdAck::Error(err.into())
+    }
+
+    #[must_use = "if you intended to assert that this is a success, consider `.unwrap()` instead"]
+    #[inline]
+    pub const fn is_success(&self) -> bool {
+        matches!(*self, StdAck::Success(_))
+    }
+
+    #[must_use = "if you intended to assert that this is an error, consider `.unwrap_err()` instead"]
+    #[inline]
+    pub const fn is_error(&self) -> bool {
+        !self.is_success()
+    }
+
+    /// Serialized the ack to binary using JSON. This used for setting the acknowledgement
+    /// field in IbcReceiveResponse.
+    ///
+    /// ## Examples
+    ///
+    /// Show how the acknowledgement looks on the write:
+    ///
+    /// ```
+    /// # use cosmwasm_std::StdAck;
+    /// let ack1 = StdAck::success(b"\x01"); // 0x01 is a FungibleTokenPacketSuccess from ICS-20.
+    /// assert_eq!(ack1.to_binary(), br#"{"result":"AQ=="}"#);
+    ///
+    /// let ack2 = StdAck::error("kaputt"); // Some free text error message
+    /// assert_eq!(ack2.to_binary(), br#"{"error":"kaputt"}"#);
+    /// ```
+    ///
+    /// Set acknowledgement field in `IbcReceiveResponse`:
+    ///
+    /// ```
+    /// use cosmwasm_std::{StdAck, IbcReceiveResponse};
+    ///
+    /// let ack = StdAck::success(b"\x01"); // 0x01 is a FungibleTokenPacketSuccess from ICS-20.
+    ///
+    /// let res: IbcReceiveResponse = IbcReceiveResponse::new().set_ack(ack.to_binary());
+    /// let res: IbcReceiveResponse = IbcReceiveResponse::new().set_ack(ack); // Does the same but consumes the instance
+    /// ```
+    pub fn to_binary(&self) -> Binary {
+        // We need a non-failing StdAck -> Binary conversion to allow using StdAck in
+        // `impl Into<Binary>` arguments.
+        // Pretty sure this cannot fail. If that changes we can create a non-failing implementation here.
+        to_binary(&self).unwrap()
+    }
+
+    pub fn unwrap(self) -> Binary {
+        match self {
+            StdAck::Success(data) => data,
+            StdAck::Error(err) => panic!("{}", err),
+        }
+    }
+
+    pub fn unwrap_err(self) -> String {
+        match self {
+            StdAck::Success(_) => panic!("not an error"),
+            StdAck::Error(err) => err,
+        }
+    }
+}
+
+impl From<StdAck> for Binary {
+    fn from(original: StdAck) -> Binary {
+        original.to_binary()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn stdack_success_works() {
+        let success = StdAck::success(b"foo");
+        match success {
+            StdAck::Success(data) => assert_eq!(data, b"foo"),
+            StdAck::Error(_err) => panic!("must not be an error"),
+        }
```

Comment:
Not a fan of `assert!(matches!(...))`?

Reply:
Does this allow me to compare the `data` value too?

Reply:
Yes, like this: `assert!(matches!(success, StdAck::Success(data) if data == b"foo"))`